api/src/services/bulk-docs.js

```diff
@@ -107,6 +111,127 @@ const deleteDocs = (docsToDelete, deletionAttemptCounts = {}, updateAttemptCount
     });
 };
 
+const filterNewDocs = (allowedDocIds, docs) => {
+  let docIds = _.unique(_.compact(docs.map(doc => doc._id)));
+
+  if (!docIds.length) {
+    // all docs are new
+    return Promise.resolve(docs);
+  }
+
+  docIds = _.difference(docIds, allowedDocIds);
+
+  if (!docIds.length) {
+    // all docs are allowed
+    return Promise.resolve([]);
+  }
+
+  return db.medic
+    .allDocs({ keys: docIds })
+    .then(result => {
+      return docs.filter(doc => {
+        // return docs without ids or docs which do not exist
+        return !doc._id || (allowedDocIds.indexOf(doc._id) === -1 && !result.rows.find(row => row.id === doc._id));
+      });
+    });
+};
+
+// iterates over request `docs`, filtering the ones the user is allowed to see or create
+// when a doc adds new `subjectIds`, the remaining `docs` are reiterated
+const filterAllowedDocs = (authorizationContext, docs) => {
+  const allowedDocs = [];
+  let shouldCheck = true,
+      pendingDocs = docs.map(doc => ({
+        doc,
+        viewResults: authorization.getViewResults(doc),
+        alwaysAllowCreate: authorization.alwaysAllowCreate(doc)
+      }));
+
+  const checkDoc = (docObj) => {
+    const allowed = docObj.alwaysAllowCreate ||
+                    authorization.allowedDoc(docObj.doc._id, authorizationContext, docObj.viewResults);
+
+    if (!allowed) {
+      return;
+    }
+
+    shouldCheck = allowed.newSubjects;
+    allowedDocs.push(docObj.doc);
+    pendingDocs = _.without(pendingDocs, docObj);
+  };
+
+  while (pendingDocs.length && shouldCheck) {
+    shouldCheck = false;
+    pendingDocs.forEach(checkDoc);
+  }
```

Comment:
This is quite a complicated little loop. It looks like you're re-running `checkDoc` in case the doc is allowed on the second call, but it's not clear why it would be allowed the second time but not the first. The three parameters passed to `allowedDoc` all appear to be unchanged between runs. Is there some side-effect here I'm not seeing? Can you think of any way to make this easier to read?

Reply:
`allowedDoc` mutates `authorizationContext.subjectIds` when new allowed contacts are tested. If the docs are out of 'logical' order (fe: a report about a new contact is before the actual contact in the array), we'd need a second pass to catch the stray reports with the correct list of allowed subjectIds.

Reply:
Do you think iterating twice explicitly is better? First pass to only check `contacts` and second pass to check everything else?

Reply:
It's not at all clear that `allowedDoc` mutates the context so that makes it hard to read. I think it might be worth having two functions - one called `updateAuthorizationContext` and one called `allowedDoc` (or whatever names you prefer). It also appears to me that the mutation only has to happen once for this loop so the `updateAuthorizationContext` call can be invoked before the loop executes.

It's also really complicated that `allowedDoc` returns different types. I'm wondering if this could be part of the new `updateAuthorizationContext` function so `allowedDoc` only returns a boolean.

Finally this function is very similar to the `processPendingChanges` function in the `changes` controller - can we reuse that function in some way (eg: by pulling it into the `authorization` service)?